[Scan to Update Inventory] Inventory Screen UI

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11307 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
With this PR we implement the UI of the update inventory screen:

- Layout
- Strictly layout logic: when showing the keyboard the first item, we change the button title. Anytime the keyboard is on screen, the image is hidden
- View model:  populate the layout with the inventory item. Here the challenge was to abstract the screen for products and variations, which have a slightly different structure.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Go to products
2. Scan a product whose inventory can be managed (`manageStock` property to true). The Inventory Update screen should appear.
3. Tap on the quantity, and make sure that the button title changes and the image disappears while the keyboard is onscreen.

Repeat with a product variation. See that the name is updated after a while (we have to retrieve the parent's product name)

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/1864060/3e72e4c7-61ba-44ca-9809-9703d5949e55

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

toupper committed:
UI Design of the UpdateProductInventoryView

toupper committed:
Fill view with product or variation

toupper committed:
Improvements on logic

toupper committed:
Add unit tests

toupper committed:
Remove accessibility identifiers (for now) and unnecessary self

toupper committed:
Add comment

toupper committed:
Tweak design

joshheald reviewed (commented):
Still to review the code, but a few thoughts on the functionality since this is WIP...

1. Variations having "Product name" while waiting for the load of the real name looks odd. Perhaps it should be a redacted/shimmering placeholder instead?
2. When it loads, the parent product name feels insufficient. If we had the variation name on the screen as well, the merchant could be more confident they're updating the right product, even when the image is hidden because the keyboard is visible.
3. On many phones, there seems to be enough space to still show the product image, even when the keyboard, buttons, description, quantity field. Perhaps making it resizable would be enough?
4. It didn't feel obvious to me that the quantity was an editable text field, when I first opened the screen. I thought my only option was to increase the quantity by +1. Perhaps some different styling for the text field would help, or just default to having it focused?

I'll review the code now

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -286,13 +287,21 @@ private extension ProductsViewController {
 
         let productSKUBarcodeScannerCoordinator = ProductSKUBarcodeScannerCoordinator(sourceNavigationController: navigationController,
                                                                                       onSKUBarcodeScanned: { [weak self] scannedBarcode in
+            guard let self = self else { return }
```
```suggestion
            guard let self else { return }
```
Just small, but no need for the `= self` any more

toupper replied:
Good suggestion! Done in 6eef8adf0444577088efe5e17bf9c180da2d4235

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -286,13 +287,21 @@ private extension ProductsViewController {
+                do {
+                    let scannedItem = try await self.viewModel.handleScannedBarcode(scannedBarcode)
+                    let viewModel = UpdateProductInventoryViewModel(inventoryItem: scannedItem.inventoryItem, siteID: self.viewModel.siteID)
+                    self.present(UIHostingController(rootView: UpdateProductInventoryView(viewModel: viewModel)), animated: true)
```
Consider passing the two parameters to the view, and having it create its viewModel. It can make the lifecycle easier to reason about. 

If you do that, and create the view model when `body` is called, it won't be created until the view is shown. It won't make much difference here at the moment, because UIKit's controlling the presentation so this happens when we are about to present it, but it could make the view more resilient for future presentation from SwiftUI.

joshheald commented on WooCommerce/Classes/ViewRelated/Products/UpdateInventory/UpdateProductInventoryView.swift:
```diff
@@ -0,0 +1,119 @@
+                    .onReceive(Publishers.keyboardHeight) { keyboardHeight in
+                        isKeyboardVisible = keyboardHeight > 0
+                    }
+                    .frame(width: geometry.size.width)
```
Combine the two frame modifiers, or at least put them in the same place?

toupper replied:
Good idea, I move them together in e3ab92bbd63f373a520c68e59421680ff22ceca9

joshheald commented on WooCommerce/Classes/ViewRelated/Products/UpdateInventory/UpdateProductInventoryView.swift:
```diff
@@ -0,0 +1,119 @@
+    /// Scale of the view based on accessibility changes
+    @ScaledMetric private var scale: CGFloat = 1.0
+
+    @Environment(\.presentationMode) var presentationMode
```
```suggestion
    @Environment(\.dismiss) private var dismiss
```

Then you can use `dismiss()` later on in the button. `presentationMode` was needed before iOS 15, but now we can use the clearer version, if it's just for dismissal.

toupper replied:
Yeah, done in 019fe664fe67366c8596d9e8dde08ed57837989c

joshheald commented on WooCommerce/Classes/ViewRelated/Products/UpdateInventory/UpdateProductInventoryViewModel.swift:
```diff
@@ -0,0 +1,85 @@
+extension ProductVariation: InventoryItem {
+    func retrieveName(with stores: StoresManager, siteID: Int64) async throws -> String {
+        // Let's retrieve the parent product's name
+        return try await withCheckedThrowingContinuation { continuation in
+            let action = ProductAction.retrieveProduct(siteID: siteID,
+                                                       productID: productID) { result in
+                switch result {
+                case let .success(product):
+                    continuation.resume(with: .success(product.name))
+                case let .failure(error):
+                    continuation.resume(throwing: error)
+                }
+            }
+
+            Task { @MainActor in
+                stores.dispatch(action)
+            }
+        }
```
Using checked continuations to call Actions in the WooCommerce target can make testing impossible, in my experience. 

I found a decent workaround, similar to what they've done on Android; create a super simple Service in Yosemite which wraps the actions in an async function. That wrapper still won't be testable itself, but the WooCommerce functions which call it can be. 

See `SystemStatusService` for an example...

toupper replied:
Thanks for the suggestion, I remember that case. I will follow the tests run on CI and adopt your strategy if they fail 👍

toupper committed:
Show loading effect when the name is loading

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Products/UpdateInventory/UpdateProductInventoryViewModelTests.swift:
```diff
@@ -0,0 +1,107 @@
+        let name = "test-name"
+        let parentProduct = Product.fake().copy(siteID: siteID, productID: parentProductID, name: name)
+        let stores = MockStoresManager(sessionManager: .makeForTesting())
+        stores.whenReceivingAction(ofType: ProductAction.self) { [weak self ]action in
+            guard let self = self else { return }
+            switch action {
+            case let .retrieveProduct(passingSiteID, productID, onCompletion):
+                if passingSiteID == self.siteID,
+                   productID == parentProductID {
+                    onCompletion(.success(parentProduct))
+                }
+            default:
+                break
+            }
+        }
+
+        let product = ProductVariation.fake().copy(siteID: siteID, productID: parentProductID)
+        let viewModel = UpdateProductInventoryViewModel(inventoryItem: product, siteID: siteID, stores: stores)
+
+        waitUntil {
+            viewModel.name == name
+        }
+    }
```
This is the kind of test I had difficulty getting to work. It would appear to work in Xcode, but then fail in CI. Really it was failing locally as well, but not reporting that failure. The test runner just couldn't figure out the different versions of async in play.

Your test does seem to be unaffected by these issues, but I'm still a bit skeptical because of my previous frustration with this. Perhaps it's some difference between `waitFor` (which I was using) and `waitUntil`.

joshheald reviewed (approved):
Works well, and the code looks good. A few suggestions, alongside those in my earlier comment, let me know what you think.

toupper committed:
Show a different button depending on the action

toupper committed:
Add border to textfield to make obvious that it can be edited

toupper committed:
Enable button depending on whether the quantity can be submitted or not

toupper committed:
Remove unused string

toupper commented:
Thanks for your comments and suggestions @joshheald! Some of them I wanted to address in following PRs, but now you mentioned them, I tackled them here:

> Variations having "Product name" while waiting for the load of the real name looks odd. Perhaps it should be a redacted/shimmering placeholder instead?

Indeed, I added a `ProgressView` while loading:


https://github.com/woocommerce/woocommerce-ios/assets/1864060/9893528b-7184-4f2b-8033-141e8abf453b

> When it loads, the parent product name feels insufficient. If we had the variation name on the screen as well, the merchant could be more confident they're updating the right product, even when the image is hidden because the keyboard is visible.

Good remark! I also considered the same. The problem is that a variation doesn't have a name per se

<img width="1166" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/7315fe9c-8778-425c-af6c-b62d9293fd2a">

We normally enumerate their attributes to identify them, but in this case, I thought it wasn't necessary because:

- The user is scanning the variation, so they should be able to identify it.
- We show the SKU
- We show the picture.

In any case, I will ask about that and align the behavior with Android.

> On many phones, there seems to be enough space to still show the product image, even when the keyboard, buttons, description, quantity field. Perhaps making it resizable would be enough?

I will ask that to the designer and project lead as well 👍 

> It didn't feel obvious to me that the quantity was an editable text field, when I first opened the screen. I thought my only option was to increase the quantity by +1. Perhaps some different styling for the text field would help, or just default to having it focused?

True, I added a border to the text field so it can be more obvious that it's editable:
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/946f4614-7396-433c-89ae-a5ad51de29aa" width="375">

toupper committed:
Make the guard shorter

toupper committed:
Remove not necessary string

toupper committed:
Simplify view - view model creation

toupper committed:
Use dismiss directly instead of the presentation mode

toupper committed:
Move frame modifiers together

toupper committed:
Move the padding modifier so the frame can be captured correctly
